flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/join/TimeBoundedStreamInnerJoin.scala

```diff
@@ -131,340 +116,308 @@ class TimeBoundedStreamInnerJoin(
+    val oppositeUpperBound: Long =  rowTime + rightRelativeSize
     processElement(
       cRowValue,
-      timeForRecord,
+      rowTime,
       ctx,
       out,
       rightOperatorTime,
+      oppositeLowerBound,
+      oppositeUpperBound,
       leftOperatorTime,
       leftTimerState,
       rightCache,
       leftCache,
-      false
+      leftRow = false
     )
   }
 
   /**
-    * Put a record from the input stream into the cache and iterate the opposite cache to
-    * output records meeting the join conditions. If there is no timer set for the OPPOSITE
+    * Put a row from the input stream into the cache and iterate the opposite cache to
+    * output join results meeting the conditions. If there is no timer set for the OPPOSITE
     * STREAM, register one.
     */
   private def processElement(
-    cRowValue: CRow,
-    timeForRecord: Long,
-    ctx: CoProcessFunction[CRow, CRow, CRow]#Context,
-    out: Collector[CRow],
-    myWatermark: Long,
-    oppositeWatermark: Long,
-    oppositeTimeState: ValueState[Long],
-    recordListCache: MapState[Long, JList[Row]],
-    oppositeCache: MapState[Long, JList[Row]],
-    leftRecord: Boolean): Unit = {
-    if (relativeWindowSize > 0) {
-      //TODO Shall we consider adding a method for initialization with the context and collector?
-      cRowWrapper.out = out
-
-      val record = cRowValue.row
-
-      //TODO Only if the time of the record is greater than the watermark, can we continue.
-      if (timeForRecord >= myWatermark - allowedLateness) {
-        val oppositeLowerBound: Long =
-          if (leftRecord) timeForRecord - rightRelativeSize else timeForRecord - leftRelativeSize
-
-        val oppositeUpperBound: Long =
-          if (leftRecord) timeForRecord + leftRelativeSize else timeForRecord + rightRelativeSize
-
-        // Put the record into the cache for later use.
-        val recordList = if (recordListCache.contains(timeForRecord)) {
-          recordListCache.get(timeForRecord)
-        } else {
-          new util.ArrayList[Row]()
-        }
-        recordList.add(record)
-        recordListCache.put(timeForRecord, recordList)
-
-        // Register a timer on THE OTHER STREAM to remove records from the cache once they are
-        // expired.
-        if (oppositeTimeState.value == 0) {
-          registerCleanUpTimer(
-            ctx, timeForRecord, oppositeWatermark, oppositeTimeState, leftRecord, true)
-        }
+      cRowValue: CRow,
+      timeForRow: Long,
+      ctx: CoProcessFunction[CRow, CRow, CRow]#Context,
+      out: Collector[CRow],
+      myWatermark: Long,
+      oppositeLowerBound: Long,
+      oppositeUpperBound: Long,
+      oppositeWatermark: Long,
+      oppositeTimeState: ValueState[Long],
+      rowListCache: MapState[Long, JList[Row]],
+      oppositeCache: MapState[Long, JList[Row]],
+      leftRow: Boolean): Unit = {
+    cRowWrapper.out = out
+    val row = cRowValue.row
+    if (!checkRowOutOfDate(timeForRow, myWatermark)) {
```

Comment:
Hi @fhueske, it takes me a little time to rethink these conditions. As you said, it's really quite easy to get confused 😄 

In general, I'm in favour of your suggestions. Besides, I've got some extra ideas. 
1. About storing a record in state. Only if `oppositeOperatorTime < oppositeQualifedUpperBound`, can a record be stored. For example, suppose there's a record from the left stream with `time = 10` and qualified time range of the right stream is calculated to be `[8, 15]`. Then only if the `rightOperatorTime < 15`, need we cache the left record. Otherwise, it means all the qualified records from the right stream should have been seen.
2. We need two conditions here, but the "lateness condition" should be applied out of the "storing condition", i.e., we first check the lateness of a record and only if it passes can we continue processing it (checking if it should be stored, etc.). That's because even if there exist records from the opposite stream that can be joined with a late record, we can not guarantee the **completeness** of the results. In the above example, if `rightOperatorTime = 16`, the left record with `time = 10` will not be stored. After that, here comes a right record with `time = 15`, but we can never join it with the left record just seen before since it is not stored.

"Incomplete" or "empty", which one to choose?

Reply:
I think we should try to produce results which are as complete as possible. In order to avoid the situation that you described in the 2. case, I suggested to include `allowedLateness` into the condition that checks if an element should be stored. So the operator time of the other input needs to be adjusted by that.

I think the storing and joining conditions are independent of each other. So I would do something like:

```
if (upperWindowBound > otherOpTime - allowedLateness) { // or >= ?
  // store record
}
if (myRecord > myOpTime - allowedLateness) { // or >= ?
  // join record
}
```

Reply:
Yes, we must include the `allowedLateness` into all the conditions. However, I think the "storing condition" should be covered by the "lateness condition", i.e.,
```
if (myRecord > myOpTime - allowedLateness) {
    if (upperWindowBound > otherOpTime - allowedLateness) {
        // store record
    }
  // join record
}
```
That's because once we store a record `R` which doesn't pass the lateness check, it may be joined with a later coming record form the opposite stream. Now that `R` isn't joined when it comes, to keep the completeness, it should never be joined.

Reply:
Thanks for your thoughts. I understand your motivation to either emit complete results for a row or none but I think it is fine to emit partial results. My reasoning is the following:

1. we should try to produce results which are as close to the original semantics as possible
2. being late (later than the allowedLateness) does not necessarily mean that we missed matching results. Maybe there were no matching records in the matched time frame, so we could still produce exact results.
3. By not storing a row in the cache, the join result becomes more incomplete. Rows of the other stream that would have produced a complete result will only produce partial results (which we wanted to avoid by not adding it to state).

So, I'd rather unnest the conditions.

Reply:
Fine. Let's produce as many results as we can!

Reply:
Hi @fhueske, I got another concern that if we produce partial results, will that affect the  batch/stream SQL consolidation? Besides, the semantics of the watermark will not be hold there. I suggest to consult more people and make this decision carefully.

Reply:
There is no problem with the unified semantics for batch and streaming. In fact, handling as much late data as possible will result in closer results of streaming and batch queries. 

You can think of a batch table as a streaming table that does not provide watermark except for one `Long.MAX_VALUE` watermark after all rows have been shipped. Hence, batch tables have no late data and the more (late) data a streaming query processes the closer its result is to the result of an equivalent batch query.

Reply:
Well, that makes sense to me. To achieve it, I think we should compare the **real lowest timestamp** in the opposite cache with the `oppositeQualifiedUpperBound`, i.e., 
```
if (upperWindowBound > otherOpTime - allowedLateness) {
  // store record
}
if (upperWindowBound > realLowestTimestamp) {
  // join record
}
```
I'll update the implementation and the tests.